jupyter_server_proxy/manager.py

```diff
@@ -0,0 +1,127 @@
+
+        # Add proxy server metadata
+        self.server_proxy_apps.append(
+            ServerProxy(
+                name=name,
+                url=ujoin(base_url, name),
+                cmd=" ".join(cmd),
+                port=port,
+                managed=True if proc else False,
+            )
+        )
+
+        # Add proxy server proc object so that we can send SIGTERM
+        # when user chooses to shut it down
+        self._server_proxy_procs.append(
+            ServerProxyProc(
+                name=name,
+                proc=proc,
+            )
+        )
+
+    async def del_server_proxy_app(self, name):
+        """Remove a launched proxy server from list"""
+        self.server_proxy_apps = [
+            app for app in self.server_proxy_apps if app.name != name
+        ]
+        self._server_proxy_procs = [
+            app for app in self._server_proxy_procs if app.name != name
+        ]
+        self.num_active_server_proxy_apps -= 1
+
+    def get_server_proxy_app(self, name):
+        """Get a given server proxy app"""
+        return next((app for app in self.server_proxy_apps if app.name == name), {})
+
+    def _get_server_proxy_proc(self, name):
+        """Get a given server proxy app"""
+        return next((app for app in self._server_proxy_procs if app.name == name), {})
+
+    def list_server_proxy_apps(self):
+        """List all active server proxy apps"""
+        return self.server_proxy_apps
+
+    def _list_server_proxy_procs(self):
+        """List all active server proxy proc objs"""
+        return self._server_proxy_procs
+
+    async def terminate_server_proxy_app(self, name):
+        """Terminate a server proxy by sending SIGTERM"""
+        app = self._get_server_proxy_proc(name)
+        try:
+            # Here we send SIGTERM signal to terminate proxy app
+            # graciously so we can restart it if needed. Note that
+            # some servers may not get stopped by sending SIGTERM
+            # signal (example is mlflow server). In this case, it is
+            # user's responsibility to write wrapper scripts around
+            # proxy app's executable to terminate them cleanly using
+            # TERM signal. It is also important to set exit code to 0
+            # when using such wrappers when proxy apps shutdown.
+            await app.proc.terminate()
+
+            # Remove proxy app from list
+            await self.del_server_proxy_app(name)
+
+            return True
+        except (KeyError, AttributeError):
+            return None
+
+    async def terminate_all(self):
+        """Close all server proxy and cleanup"""
+        for app in self.server_proxy_apps:
+            await self.terminate_server_proxy_app(app)
+
+
+# Create a default manager to keep track of server proxy apps.
+manager = ServerProxyAppManager()
+
+# Create a Periodic call back function to check the status of processes
+pc = PeriodicCallback(monitor_server_proxy_procs, 1e4)
+pc.start()
```

Comment:
This could be reasonable, but I'm not confident on it. Is it acceptable that we start this up automatically as soon as this file is loaded?

I'm not confident on the alternate options, but I figure this is a singleton instance of the class defined above, created when this python file is loaded. I figure options involve delaying this initialization.

I'm also thinking a bit about if `monitor_server_proxy_procs` should or shouldn't be part of the class itself. It is a function tied to a specific object of the class anyhow via the `manager` variable that could be `self` as well.

If not someone has a concrete change idea here to persue or generally considers this to be blocking, I don't think this comment should be blocking a merge.

Reply:
Right. Importing this file shouldn't have side-effects like starting/changing processes or even things on the event loop, as there's no telling it will actually be there.

Ideally, as much business logic/exception handling as possible would move into the manager class, and out of the handlers and `__init__.py`. The server extension would then become a very small layer that just hands down the parent application into the manager.

Reply:
Also, I'd recommend new-start async stuff to use stdlib `asyncio` or  [`anyio`](https://anyio.readthedocs.io/en/stable/) primitives instead of `tornado`-specific ones. And indeed, if we could get _out_ of the polling business, that would be ideal: the underlying processes probably do (or should) provide callbacks/events that can be used to directly update the manager state.

Assuming we _do_ have to poll, for some reason, and as I don't actually see a drop-in for `PeriodicCallback`, this could be a `ServerProxyManager.monitor()` which fires off an infinite loop:

```python
import traitlets
import asyncio 

class ServerProxyManager(ServerProxy):
    monitor_interval = traitlets.Int(10, help="Proxy polling interval in seconds").tag(config=True)
    monitoring = traitlets.Bool(False, help="Whether the manager should poll for proxy status")

    # this doesn't need to be a trait
    _monitor_task: asyncio.Task

    @traitlets.observe("monitoring")
    def on_monitoring_changed(self, change: traitlets.Bunch):
        if self._monitor_task:
            self._monitoring_task.cancel()
        if change.new:
            # keep a handle so it doesn't get GCd
            self._monitoring_task = asyncio.create_task(self._monitor_loop()) 

    async def _monitor_loop(self):
        while self.monitoring:
             # do the thing
             await asyncio.sleep(self.monitor_interval)
```

When this gets invoked from `__init__.py`:

```
def _load_jupyter_server_extension(nbapp: ServerApp):
    manager = ServerProxyManager(parent=nbapp)
    manager.monitor()
```

Reply:
If a user gracefully terminates a server proxy _manually_ outside of lab interface, that process will not be restarted as we are not using [`always_restart`](https://github.com/jupyterhub/simpervisor/blob/425c81f56c76756e3a02cb9f9a108d543964324d/simpervisor/process.py#L134) feature of simpervisor. So somehow we need to update manager to remove this server proxy app and that is why I included this polling. Simpervisor uses exit handlers internally but I think they are not configurable. So, I dont know how we can fire a callback when process exits gracefully.